### PR TITLE
Type Providers: fix races in TypeProvidersManager & ExtensionTypingProviderShim

### PR DESCRIPTION
**Pull request by DedSec256**
Fixed cases:

1. Concurrent unsynchronized creation of the TypeProvidersConnection, leading to data loss
2. Concurrent creation  of type providers would cause the requested type provider to be missing from the  in-process TypeProvidersCache

### CODE REVIEW

File: ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/ExtensionTypingProviderShim.fs [hunk 1]
```diff
@@ -26,8 +26,9 @@ type ExtensionTypingProviderShim(solution: ISolution, toolset: ISolutionToolset,
     let lifetime = solution.GetLifetime()
     let defaultShim = ExtensionTypingProvider
     let outOfProcess = experimentalFeatures.OutOfProcessTypeProviders
+    let createProcessLockObj = obj()
 
-    let mutable connection: TypeProvidersConnection = null
+    let [<VolatileField>] mutable connection: TypeProvidersConnection = null
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/ExtensionTypingProviderShim.fs:31]: Is there any evidence this change is needed?

File: ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/ExtensionTypingProviderShim.fs [hunk 2]
```diff
@@ -38,10 +39,15 @@ type ExtensionTypingProviderShim(solution: ISolution, toolset: ISolutionToolset,
...
-            typeProvidersManager <- TypeProvidersManager(connection) :?> _
+        if isConnectionAlive () then () else
+
+        lock createProcessLockObj (fun () ->
+
+        if isConnectionAlive () then () else
+
+        typeProvidersHostLifetime <- Lifetime.Define(lifetime)
+        connection <- typeProvidersLoadersFactory.Create(typeProvidersHostLifetime.Lifetime).Run()
+        typeProvidersManager <- TypeProvidersManager(connection) :?> _)
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/ExtensionTypingProviderShim.fs:50]: Please indent lambda body.

File: ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/TypeProvidersManager.fs [hunk 3]
```diff
@@ -27,26 +28,29 @@ type internal TypeProvidersCache() =
...
         if typeProvidersPerAssembly.[envKey].Count = 0 then
-            typeProvidersPerAssembly.Remove(envKey) |> ignore
+            typeProvidersPerAssembly.TryRemove(envKey) |> ignore
 
     member x.Add(envKey, tp) =
         addTypeProvider envKey tp
 
     member x.Get(id) =
-        proxyTypeProvidersPerId.[id]
+        let hasValue = SpinWait.SpinUntil((fun () -> proxyTypeProvidersPerId.ContainsKey id), 15_000)
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/TypeProvidersManager.fs:50]: Where does `15_000` literal comes from?

Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/TypeProvidersManager.fs:50]: this is a fairly random number (15 seconds). It is assumed that the data will appear in the cache quickly enough, but for an unforeseen situation, we can wait a little longer.

File: ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/ExtensionTypingProviderShim.fs (hunk 1 above)
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/ExtensionTypingProviderShim.fs:31]: "Does this need to be volatile? Best answer is Yes." (c)

The common implementation of double-checking locking requires a volatile attribute on the field being initialized, especially if the code works on different architectures.

File: ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/ExtensionTypingProviderShim.fs (hunk 2 above)
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/ExtensionTypingProviderShim.fs:50]: oh, a situation is possible when the connection is initialized, but the manager is not

File: ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/TypeProvidersManager.fs (hunk 3 above)
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/TypeProvidersManager.fs:50]: do you think 15 seconds is a lot?

Comment (auduchinok) [reply ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/TypeProvidersManager.fs:50]: No, it's choosing the time arbitrary seems very suspicious to me.

File: ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/ExtensionTypingProviderShim.fs (hunk 1 above)
Comment (auduchinok) [reply ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/ExtensionTypingProviderShim.fs:31]: Shouldn't it be added to Fantomas host connection too? What about other connections in the product?

**Pull request merged**